troubleshooting/checkpointing/src/solution/java/com/ververica/flink/training/solutions/CheckpointingJobSolution3.java

```diff
@@ -0,0 +1,300 @@
+     */
+	public static void main(String[] args) throws Exception {
+		ParameterTool parameters = ParameterTool.fromArgs(args);
+
+		StreamExecutionEnvironment env = createConfiguredEnvironment(parameters);
+
+		//Timing Configuration
+		env.getConfig().setAutoWatermarkInterval(100);
+		env.setBufferTimeout(10);
+
+		//Checkpointing Configuration (use cluster-configs if not run locally)
+		if (isLocal(parameters)) {
+			env.enableCheckpointing(TimeUnit.SECONDS.toMillis(10));
+			env.getCheckpointConfig().setMinPauseBetweenCheckpoints(TimeUnit.SECONDS.toMillis(10));
+			env.getCheckpointConfig().setCheckpointTimeout(TimeUnit.MINUTES.toMillis(2));
+		}
+
+		DataStream<Tuple2<Measurement, Long>> sourceStream = env
+				.addSource(SourceUtils.createFailureFreeFakeKafkaSource())
+				.name("FakeKafkaSource")
+				.uid("FakeKafkaSource")
+				.assignTimestampsAndWatermarks(
+						WatermarkStrategy
+								.<FakeKafkaRecord>forBoundedOutOfOrderness(Duration.ofMillis(250))
+								.withTimestampAssigner(
+										(element, timestamp) -> element.getTimestamp())
+								.withIdleness(Duration.ofSeconds(1)))
+				.name("Watermarks")
+				.uid("Watermarks")
+				.flatMap(new MeasurementDeserializer())
+				.name("Deserialization")
+				.uid("Deserialization");
+
+		DataStream<Tuple2<Measurement, Long>> sortedStream = sourceStream
+				.keyBy(x -> x.f0.getSensorId())
+				.process(new SortMeasurementFunction())
+				.name("Sorting")
+				.uid("Sorting");
+
+		KeyedStream<Tuple2<Measurement, Long>, Integer> keyedSortedStream =
+				DataStreamUtils.reinterpretAsKeyedStream(
+						sortedStream,
+						x -> x.f0.getSensorId());
+
+		DataStream<WindowedMeasurements> aggregatedPerLocation = keyedSortedStream
+				.window(SlidingEventTimeWindows.of(Time.of(1, TimeUnit.MINUTES), Time.of(1, TimeUnit.SECONDS)))
+				.aggregate(new MeasurementWindowAggregatingFunction(),
+						new MeasurementWindowProcessFunction())
+				.name("WindowedAggregationPerLocation")
+				.uid("WindowedAggregationPerLocation");
+
+		if (isLocal(parameters)) {
+			aggregatedPerLocation.print()
+					.name("NormalOutput")
+					.uid("NormalOutput")
+					.disableChaining();
+		} else {
+			aggregatedPerLocation.addSink(new DiscardingSink<>())
+					.name("NormalOutput")
+					.uid("NormalOutput")
+					.disableChaining();
+		}
+
+		env.execute(CheckpointingJobSolution3.class.getSimpleName());
+	}
+
+	public static class SortMeasurementFunction
+			extends KeyedProcessFunction<Integer, Tuple2<Measurement, Long>, Tuple2<Measurement, Long>> {
+
+		private ListState<Tuple2<Measurement, Long>> listState;
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
+			super.open(parameters);
+
+			ListStateDescriptor<Tuple2<Measurement, Long>> desc =
+					new ListStateDescriptor<Tuple2<Measurement, Long>>(
+							"events",
+							TypeInformation.of(new TypeHint<Tuple2<Measurement, Long>>(){})
+					);
```

Comment:
Do you think, the following would be more concise?
```suggestion
					new ListStateDescriptor<>(
							"events",
							Types.TUPLE(Types.POJO(Measurement.class), Types.LONG)
					);
```